src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c

```diff
@@ -0,0 +1,945 @@
+
+	if (zdev->zdev_xclbin->zx_refcnt == 0)
+		DRM_INFO("now xclbin can be changed");
+
+	DRM_INFO("<- Release xclbin %pUB, to ref=%d",
+	    xclbin_uuid, zdev->zdev_xclbin->zx_refcnt);
+
+	return 0;
+}
+
+int
+zocl_xclbin_ctx(struct drm_zocl_dev *zdev, struct drm_zocl_ctx *ctx,
+	struct sched_client_ctx *client)
+{
+	struct sched_exec_core *exec = zdev->exec;
+	xuid_t *zdev_xuid, *ctx_xuid = NULL;
+	u32 cu_idx = ctx->cu_index;
+	int ret = 0;
+
+	BUG_ON(!mutex_is_locked(&zdev->zdev_xclbin_lock));
+
+	ctx_xuid = vmalloc(ctx->uuid_size);
+	if (!ctx_xuid)
+		return -ENOMEM;
+	ret = copy_from_user(ctx_xuid, (void *)ctx->uuid_ptr, ctx->uuid_size);
+	if (ret) {
+		vfree(ctx_xuid);
+		return ret;
+	}
+
+	/*
+	 * valid xclbin_id is the same.
+	 * Note: xclbin has been downloaded by read_axlf.
+	 *       user can only open/remove context with same loaded xclbin.
+	 */
+	zdev_xuid = (xuid_t *)zdev->zdev_xclbin->zx_uuid;
+
+	if (!zdev_xuid || !uuid_equal(zdev_xuid, ctx_xuid)) {
+		DRM_ERROR("try to add/remove CTX with wrong xclbin %pUB",
+		    ctx_xuid);
+		ret = -EBUSY;
+		goto out;
+	}
+
+	/* validate cu_idx */
+	if (!VIRTUAL_CU(cu_idx) && cu_idx >= zdev->ip->m_count) {
+		DRM_ERROR("CU Index(%u) >= numcus(%d)\n",
+		    cu_idx, zdev->ip->m_count);
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/* validate cu */
+	if (!VIRTUAL_CU(cu_idx) && !zocl_exec_valid_cu(exec, cu_idx)) {
+		DRM_ERROR("invalid CU(%d)", cu_idx);
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/*
+	 * handle remove or add
+	 * each client ctx can lock bitstream once, multiple ctx will
+	 * lock bitstream n times. clien is responsible releasing the refcnt
+	 */
+	if (ctx->op == ZOCL_CTX_OP_FREE_CTX) {
+		if (zocl_xclbin_refcount(zdev) == 0) {
+			DRM_ERROR("can not remove unused xclbin");
+			ret = -EINVAL;
+			goto out;
+		}
+		--client->num_cus;
+		if (CLIENT_NUM_CU_CTX(client) == 0)
+			ret = zocl_xclbin_release(zdev);
+	} else {
+		if (CLIENT_NUM_CU_CTX(client) == 0) {
+			ret = zocl_xclbin_hold(zdev, zdev_xuid);
+			if (ret)
+				goto out;
+		}
+		++client->num_cus;
+	}
+out:
+	write_unlock(&zdev->attr_rwlock);
+	vfree(ctx_xuid);
+	return ret;
+}
+
+int
+zocl_xclbin_set_uuid(struct drm_zocl_dev *zdev, void *uuid)
+{
+	xuid_t *zx_uuid = zdev->zdev_xclbin->zx_uuid;
+
+	if (zx_uuid) {
+		vfree(zx_uuid);
+		zx_uuid = NULL;
+	}
+
+	zx_uuid = vmalloc(UUID_SIZE * sizeof(u8));
+	if (!zx_uuid)
+		return -ENOMEM;
+
+	uuid_copy(zx_uuid, uuid);
+	zdev->zdev_xclbin->zx_uuid = zx_uuid;
+	return 0;
+}
+
+int
+zocl_xclbin_init(struct drm_zocl_dev *zdev)
+{
+	zdev->zdev_xclbin = vmalloc(sizeof(struct zocl_xclbin));
+	if (!zdev->zdev_xclbin)
+		return -ENOMEM;
+
+	zdev->zdev_xclbin->zx_last_bitstream = 0;
+	zdev->zdev_xclbin->zx_refcnt = 0;
+	zdev->zdev_xclbin->zx_uuid = NULL;
+
+	return 0;
+}
+void
+zocl_xclbin_fini(struct drm_zocl_dev *zdev)
+{
+	if (zdev->zdev_xclbin->zx_uuid)
+		vfree(zdev->zdev_xclbin->zx_uuid);
+	if (zdev->zdev_xclbin)
+		vfree(zdev->zdev_xclbin);
+	zdev->zdev_xclbin = NULL;
```

Comment:
In Linux, vfree() will directly return if addr is null. We don't necessary to check if addr is NULL.
This is just a coding style issue. I'm okay with this code.